[Torchelasic] Create root log directory by default

kurman opened the pull request:
Summary:
After refactoring in https://github.com/pytorch/pytorch/pull/120691, default behavior unintentionally was changes from creating tempdir for logging to not capturing any logs by torch Elastic Agent.

Reverting the behavior to:
- making tempdir when log dir is not specified
- allowing non-empty root log dir
    - Note: in case attempt folder exists, it will be pruned here: https://github.com/pytorch/pytorch/blob/main/torch/distributed/elastic/multiprocessing/api.py#L294

Differential Revision: D54531851




cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

d4l3k commented on torch/distributed/elastic/multiprocessing/api.py:
```diff
@@ -234,24 +234,19 @@ class DefaultLogsSpecs(LogsSpecs):
         if log_dir != os.devnull:
-            if not os.path.exists(log_dir):
+            if not log_dir:
+                log_dir = tempfile.mkdtemp(prefix="torchelastic_")
```
I expect this is fine in most cases -- though with this temp directory we'll never clean up the logs which could cause issues. Should we be deleting this when elastic shuts down?

kurman replied:
it was behaving this way before refactoring changes. Btw, users can now define their own logic using custom logspecs (using entrypoints)

kurman committed:
[Torchelasic] Create root log directory by default (#121257)

Summary:

After refactoring in https://github.com/pytorch/pytorch/pull/120691, default behavior unintentionally was changes from creating tempdir for logging to not capturing any logs by torch Elastic Agent.

Reverting the behavior to:
- making tempdir when log dir is not specified
- allowing non-empty root log dir
    - Note: in case attempt folder exists, it will be pruned here: https://github.com/pytorch/pytorch/blob/main/torch/distributed/elastic/multiprocessing/api.py#L294

Differential Revision: D54531851

d4l3k reviewed (approved):
LGTM

kurman commented:
@pytorchbot merge

Skylion007 commented on torch/distributed/elastic/multiprocessing/api.py:
```diff
@@ -229,36 +229,31 @@ class DefaultLogsSpecs(LogsSpecs):
-            if not os.path.exists(log_dir):
+            if not log_dir:
+                log_dir = tempfile.mkdtemp(prefix="torchelastic_")
+            elif not os.path.exists(log_dir):
                 os.makedirs(log_dir)
```
```suggestion
            os.makedirs(log_dir, exist_ok=True)
```
